File: deploy/biped_ws/src/biped_control/biped_control/policy_node.py
"""Policy inference node — 50Hz ONNX inference loop.

Subscribes to IMU, joint states, and cmd_vel. Builds 45d observation,
runs ONNX inference, converts actions to MIT commands, publishes.

Safety features:
  - Only publishes commands when state machine is STAND or WALK
  - Soft start: interpolates from current position to policy target over 2s
  - Ramps gains from 10% to full over 1s after soft start completes

Usage:
    ros2 run biped_control policy_node --ros-args \
        -p onnx_model:=student_flat.onnx \
        -p gain_scale:=0.3
"""

import time
import numpy as np
import yaml
import onnxruntime as ort
import rclpy
import rcl_interfaces.msg
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy
from sensor_msgs.msg import Imu, JointState
# JointState also used for /policy_viz_joints
from geometry_msgs.msg import Twist, Vector3Stamped
from std_msgs.msg import String
from biped_msgs.msg import MITCommand, MITCommandArray
from biped_control.obs_builder import ObsBuilder, ISAAC_JOINT_ORDER, DEFAULT_POSITIONS, DEFAULT_GAINS, ACTION_ORDER


# Soft start timing
SOFT_START_POSITION_SECS = 2.0  # interpolate current → target over this
SOFT_START_GAIN_SECS = 1.0       # ramp gains from 10% → 100% after position interp
WALK_GAIN_RAMP_SECS = 5.0        # ramp gains from 10% → 100% on WALK entry


class PolicyNode(Node):
    """50Hz RL policy inference node with soft start and state machine gating."""

    def __init__(self):
        super().__init__('policy_node')

        self.declare_parameter('onnx_model', 'student_flat.onnx')
        self.declare_parameter('loop_rate', 50.0)
        self.declare_parameter('gains_file', '')
        self.declare_parameter('gain_scale', 1.0,
            rcl_interfaces.msg.ParameterDescriptor(dynamic_typing=True))

        model_path = str(self.get_parameter('onnx_model').value)
        self._rate = float(self.get_parameter('loop_rate').value)
        gains_file = str(self.get_parameter('gains_file').value)
        self._gain_scale = float(self.get_parameter('gain_scale').value)

        # Load ONNX model
        self._session = None
        try:
            self._session = ort.InferenceSession(model_path)
            self.get_logger().info(f'ONNX model loaded: {model_path}')
        except Exception as e:
            self.get_logger().error(f'Failed to load ONNX model: {e}')

        # Load gains
        self._gains = dict(DEFAULT_GAINS)
        if gains_file:
            self._load_gains(gains_file)

        # Observation builder
        self._obs_builder = ObsBuilder()

        # Latest sensor data
        self._gyro = np.zeros(3, dtype=np.float32)
        self._gravity = np.array([0.0, 0.0, 9.81], dtype=np.float32)
        self._cmd_vel = np.zeros(3, dtype=np.float32)
        self._joint_positions = {}  # populated from /joint_states
        self._joint_velocities = {}
        self._fsm_state = "IDLE"
        self._walk_start_time = None  # set on WALK entry for gain ramp

        # QoS
        sensor_qos = QoSProfile(depth=1, reliability=ReliabilityPolicy.BEST_EFFORT)

        # Subscribers
        self.create_subscription(Imu, '/imu/data', self._imu_cb, sensor_qos)
        self.create_subscription(Vector3Stamped, '/imu/gravity', self._gravity_cb, sensor_qos)
        self.create_subscription(JointState, '/joint_states', self._joint_cb, sensor_qos)
        self.create_subscription(Twist, '/cmd_vel', self._cmd_cb, 10)
        self.create_subscription(String, '/state_machine', self._fsm_cb, 10)

        # Publishers
        self._pub_cmd = self.create_publisher(MITCommandArray, '/joint_commands', 10)
        self._pub_viz = self.create_publisher(MITCommandArray, '/policy_viz', 10)
        self._pub_viz_js = self.create_publisher(JointState, '/policy_viz_joints', 10)

        # Timer
        self._timer = self.create_timer(1.0 / self._rate, self._loop)

        self.get_logger().info(
            f'Policy node started — {self._rate}Hz, gain_scale={self._gain_scale}, '
            f'waiting for STAND/WALK state'
        )

    def _load_gains(self, path: str):
        try:
            with open(path) as f:
                data = yaml.safe_load(f) or {}
            for name in ISAAC_JOINT_ORDER:
                if name in data:
                    kp = data[name].get('kp', self._gains[name][0])
                    kd = data[name].get('kd', self._gains[name][1])
                    self._gains[name] = (kp, kd)
            self.get_logger().info(f'Loaded gains from {path}')
        except Exception as e:
            self.get_logger().warn(f'Failed to load gains: {e}, using defaults')

    # --- Callbacks ---

    def _imu_cb(self, msg: Imu):
        self._gyro[0] = msg.angular_velocity.x
        self._gyro[1] = msg.angular_velocity.y
        self._gyro[2] = msg.angular_velocity.z

    def _gravity_cb(self, msg: Vector3Stamped):
        self._gravity[0] = msg.vector.x
        self._gravity[1] = msg.vector.y
        self._gravity[2] = msg.vector.z

    def _joint_cb(self, msg: JointState):
        for i, name in enumerate(msg.name):
            if i < len(msg.position):
                self._joint_positions[name] = msg.position[i]
            if i < len(msg.velocity):
                self._joint_velocities[name] = msg.velocity[i]

    def _cmd_cb(self, msg: Twist):
        self._cmd_vel[0] = msg.linear.x
        self._cmd_vel[1] = msg.linear.y
        self._cmd_vel[2] = msg.angular.z

    def _fsm_cb(self, msg: String):
        new_state = msg.data.strip().upper()
        if new_state != self._fsm_state:
            self.get_logger().info(f'FSM: {self._fsm_state} → {new_state}')
            if new_state in ("WALK", "SIM_WALK"):
                self._walk_start_time = time.time()
            self._fsm_state = new_state

    # --- Main loop ---

    def _loop(self):
        if self._session is None:
            return

        # Run during WALK (motors) or SIM_WALK (viz only)
        if self._fsm_state not in ("WALK", "SIM_WALK"):
            return

        # Need joint state data before we can build obs
        if not self._joint_positions:
            return

        # Build observation
        obs = self._obs_builder.build(
            gyro=self._gyro,
            gravity=self._gravity,
            cmd_vel=self._cmd_vel,  # pass cmd_vel in both WALK and SIM_WALK
            joint_positions=self._joint_positions,
            joint_velocities=self._joint_velocities,
        )

        # Run inference
        obs_input = obs.reshape(1, -1).astype(np.float32)
        actions = self._session.run(None, {'obs': obs_input})[0][0]

        # Clamp actions to training range — prevents feedback divergence
        # through last_action in the observation vector
        actions = np.clip(actions, -1.0, 1.0)

        # Update last action
        self._obs_builder.update_last_action(actions)

        # Debug: comprehensive logging every 10 seconds
        if not hasattr(self, '_debug_timer'):
            self._debug_timer = 0
        self._debug_timer += 1
        if self._debug_timer % (int(self._rate) * 10) == 1:  # every 10s
            self.get_logger().info(f'[RAW] gyro={self._gyro} gravity={self._gravity} cmd_vel={self._cmd_vel}')
            self.get_logger().info(f'[RAW] joint_pos={{{", ".join(f"{k}:{v:.4f}" for k,v in sorted(self._joint_positions.items()))}}}')
            self.get_logger().info(f'[RAW] joint_vel={{{", ".join(f"{k}:{v:.4f}" for k,v in sorted(self._joint_velocities.items()))}}}')
            self.get_logger().info(f'[OBS] full_45d={np.array2string(obs, precision=4, separator=",")}')
            self.get_logger().info(f'[ACT] actions={np.array2string(actions, precision=4, separator=",")}')

        # Convert to position targets
        targets = ObsBuilder.action_to_positions(actions)

        if self._debug_timer % (int(self._rate) * 10) == 1:
            tgt_str = ' '.join(f'{n}={targets[n]:+.3f}' for n in ACTION_ORDER)
            self.get_logger().info(f'[TGT] {tgt_str}')


        # Build MIT command array
        cmd_msg = MITCommandArray()
        cmd_msg.header.stamp = self.get_clock().now().to_msg()

        # Gain ramp on WALK entry — prevent violent transition from STAND
        walk_ramp = 1.0
        if self._walk_start_time is not None:
            walk_elapsed = time.time() - self._walk_start_time
            walk_ramp = min(1.0, 0.1 + 0.9 * (walk_elapsed / WALK_GAIN_RAMP_SECS))

        for name in ISAAC_JOINT_ORDER:
            cmd = MITCommand()
            cmd.joint_name = name
            cmd.position = targets[name]
            cmd.velocity = 0.0
            kp, kd = self._gains[name]
            gs = float(self.get_parameter("gain_scale").value)
            cmd.kp = kp * gs * walk_ramp
            cmd.kd = kd * gs * walk_ramp
            cmd.torque_ff = 0.0
            cmd_msg.commands.append(cmd)

        if self._debug_timer % (int(self._rate) * 10) == 1:
            for cmd in cmd_msg.commands:
                self.get_logger().info(f'[CMD] {cmd.joint_name} pos={cmd.position:+.4f} kp={cmd.kp:.1f} kd={cmd.kd:.1f}')

        if self._fsm_state == "SIM_WALK":
            self._pub_viz.publish(cmd_msg)  # viz only, motors hold STAND
            # Also publish as JointState for Foxglove plotting
            js = JointState()
            js.header = cmd_msg.header
            js.name = [cmd.joint_name for cmd in cmd_msg.commands]
            js.position = [cmd.position for cmd in cmd_msg.commands]
            js.velocity = [0.0] * len(cmd_msg.commands)
            js.effort = [0.0] * len(cmd_msg.commands)
            self._pub_viz_js.publish(js)
        else:
            self._pub_cmd.publish(cmd_msg)  # actual motor commands


def main(args=None):
    rclpy.init(args=args)
    node = PolicyNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        try:
            rclpy.shutdown()
        except Exception:
            pass


if __name__ == '__main__':
    main()

File: deploy/biped_ws/src/biped_control_cpp/src/policy_node.cpp
#include <chrono>
#include <cmath>
#include <string>
#include <vector>
#include <unordered_map>
#include <fstream>
#include <iostream>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/imu.hpp"
#include "sensor_msgs/msg/joint_state.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "geometry_msgs/msg/vector3_stamped.hpp"
#include "std_msgs/msg/string.hpp"
#include "biped_msgs/msg/mit_command.hpp"
#include "biped_msgs/msg/mit_command_array.hpp"

#include <yaml-cpp/yaml.h>
#include <onnxruntime_cxx_api.h>
#include "biped_control_cpp/obs_builder.hpp"

using namespace std::chrono_literals;
using namespace biped_control_cpp;

// Soft start timing
const double SOFT_START_POSITION_SECS = 2.0;
const double SOFT_START_GAIN_SECS = 1.0;
const double WALK_GAIN_RAMP_SECS = 0.1;

class PolicyNodeCpp : public rclcpp::Node {
public:
    PolicyNodeCpp() : Node("policy_node") {
        declare_parameter("onnx_model", "student_flat.onnx");
        declare_parameter("loop_rate", 50.0);
        declare_parameter("gains_file", "");
        declare_parameter("gain_scale", 1.0);

        declare_parameter("control_params_file", "");
        
        std::string model_path = get_parameter("onnx_model").as_string();
        rate_ = get_parameter("loop_rate").as_double();
        std::string gains_file = get_parameter("gains_file").as_string();
        gain_scale_ = get_parameter("gain_scale").as_double();

        std::string control_params_path = get_parameter("control_params_file").as_string();
        if (!control_params_path.empty()) {
            load_control_params(control_params_path);
        } else {
            load_control_params("");
        }

        gains_ = DEFAULT_GAINS;
        load_gains(gains_file);

        // Load ONNX model
        try {
            Ort::Env env(ORT_LOGGING_LEVEL_WARNING, "PolicyNode");
            Ort::SessionOptions session_options;
            
            // Try enabling TensorRT / CUDA execution provider
            std::vector<std::string> available_providers = Ort::GetAvailableProviders();
            for (const auto& provider : available_providers) {
                if (provider == "TensorrtExecutionProvider") {
                    OrtTensorRTProviderOptions trt_options{};
                    session_options.AppendExecutionProvider_TensorRT(trt_options);
                    RCLCPP_INFO(get_logger(), "TensorRT Execution Provider enabled.");
                } else if (provider == "CUDAExecutionProvider") {
                    OrtCUDAProviderOptions cuda_options{};
                    session_options.AppendExecutionProvider_CUDA(cuda_options);
                    RCLCPP_INFO(get_logger(), "CUDA Execution Provider enabled.");
                }
            }

            session_ = std::make_unique<Ort::Session>(env, model_path.c_str(), session_options);
            
            // Get input/output info
            Ort::AllocatorWithDefaultOptions allocator;
            owned_input_names_.push_back(session_->GetInputNameAllocated(0, allocator));
            input_node_names_.push_back(owned_input_names_.back().get());

            owned_output_names_.push_back(session_->GetOutputNameAllocated(0, allocator));
            output_node_names_.push_back(owned_output_names_.back().get());

            RCLCPP_INFO(get_logger(), "ONNX model loaded: %s", model_path.c_str());
        } catch (const std::exception& e) {
            RCLCPP_ERROR(get_logger(), "Failed to load ONNX model: %s", e.what());
        }

        auto sensor_qos = rclcpp::QoS(1).best_effort();

        sub_imu_ = create_subscription<sensor_msgs::msg::Imu>(
            "/imu/data", sensor_qos, std::bind(&PolicyNodeCpp::imu_cb, this, std::placeholders::_1));
        sub_gravity_ = create_subscription<geometry_msgs::msg::Vector3Stamped>(
            "/imu/gravity", sensor_qos, std::bind(&PolicyNodeCpp::gravity_cb, this, std::placeholders::_1));
        sub_joint_ = create_subscription<sensor_msgs::msg::JointState>(
            "/joint_states", sensor_qos, std::bind(&PolicyNodeCpp::joint_cb, this, std::placeholders::_1));
        sub_cmd_ = create_subscription<geometry_msgs::msg::Twist>(
            "/cmd_vel", 10, std::bind(&PolicyNodeCpp::cmd_cb, this, std::placeholders::_1));
        sub_fsm_ = create_subscription<std_msgs::msg::String>(
            "/state_machine", 10, std::bind(&PolicyNodeCpp::fsm_cb, this, std::placeholders::_1));

        pub_cmd_ = create_publisher<biped_msgs::msg::MITCommandArray>("/joint_commands", 10);
        pub_viz_ = create_publisher<biped_msgs::msg::MITCommandArray>("/policy_viz", 10);
        pub_viz_js_ = create_publisher<sensor_msgs::msg::JointState>("/policy_viz_joints", 10);

        timer_ = create_wall_timer(
            std::chrono::duration<double>(1.0 / rate_),
            std::bind(&PolicyNodeCpp::loop_cb, this)
        );

        RCLCPP_INFO(get_logger(), "Policy node started (C++) — %.1fHz, gain_scale=%.2f", rate_, gain_scale_);
    }

private:
    double rate_, gain_scale_;
    std::unordered_map<std::string, std::pair<double, double>> gains_;
    ObsBuilder obs_builder_;

    std::array<float, 3> gyro_ = {0.0f, 0.0f, 0.0f};
    std::array<float, 3> gravity_ = {0.0f, 0.0f, 9.81f};
    std::array<float, 3> cmd_vel_ = {0.0f, 0.0f, 0.0f};
    std::unordered_map<std::string, double> joint_positions_;
    std::unordered_map<std::string, double> joint_velocities_;
    
    std::string fsm_state_ = "IDLE";
    double walk_start_time_ = -1.0;
    int debug_timer_ = 0;

    std::unique_ptr<Ort::Session> session_;
    std::vector<Ort::AllocatedStringPtr> owned_input_names_;
    std::vector<Ort::AllocatedStringPtr> owned_output_names_;
    std::vector<const char*> input_node_names_;
    std::vector<const char*> output_node_names_;

    rclcpp::Subscription<sensor_msgs::msg::Imu>::SharedPtr sub_imu_;
    rclcpp::Subscription<geometry_msgs::msg::Vector3Stamped>::SharedPtr sub_gravity_;
    rclcpp::Subscription<sensor_msgs::msg::JointState>::SharedPtr sub_joint_;
    rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr sub_cmd_;
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr sub_fsm_;

    rclcpp::Publisher<biped_msgs::msg::MITCommandArray>::SharedPtr pub_cmd_;
    rclcpp::Publisher<biped_msgs::msg::MITCommandArray>::SharedPtr pub_viz_;
    rclcpp::Publisher<sensor_msgs::msg::JointState>::SharedPtr pub_viz_js_;
    rclcpp::TimerBase::SharedPtr timer_;

    void load_gains(const std::string& path) {
        if (path.empty()) return;
        try {
            YAML::Node data = YAML::LoadFile(path);
            for (const auto& name : JOINT_ORDER) {
                if (data[name]) {
                    double kp = data[name]["kp"] ? data[name]["kp"].as<double>() : gains_[name].first;
                    double kd = data[name]["kd"] ? data[name]["kd"].as<double>() : gains_[name].second;
                    gains_[name] = {kp, kd};
                }
            }
            RCLCPP_INFO(get_logger(), "Loaded gains from %s", path.c_str());
        } catch (const std::exception& e) {
            RCLCPP_WARN(get_logger(), "Failed to load gains: %s, using defaults", e.what());
        }
    }

    void imu_cb(const sensor_msgs::msg::Imu::SharedPtr msg) {
        gyro_[0] = msg->angular_velocity.x;
        gyro_[1] = msg->angular_velocity.y;
        gyro_[2] = msg->angular_velocity.z;
    }

    void gravity_cb(const geometry_msgs::msg::Vector3Stamped::SharedPtr msg) {
        gravity_[0] = msg->vector.x;
        gravity_[1] = msg->vector.y;
        gravity_[2] = msg->vector.z;
    }

    void joint_cb(const sensor_msgs::msg::JointState::SharedPtr msg) {
        for (size_t i = 0; i < msg->name.size(); ++i) {
            if (i < msg->position.size()) joint_positions_[msg->name[i]] = msg->position[i];
            if (i < msg->velocity.size()) joint_velocities_[msg->name[i]] = msg->velocity[i];
        }
    }

    void cmd_cb(const geometry_msgs::msg::Twist::SharedPtr msg) {
        cmd_vel_[0] = msg->linear.x;
        cmd_vel_[1] = msg->linear.y;
        cmd_vel_[2] = msg->angular.z;
    }

    void fsm_cb(const std_msgs::msg::String::SharedPtr msg) {
        std::string new_state = msg->data;
        for (auto& c : new_state) c = std::toupper(c);
        
        if (new_state != fsm_state_) {
            RCLCPP_INFO(get_logger(), "FSM: %s -> %s", fsm_state_.c_str(), new_state.c_str());
            if (new_state == "WALK" || new_state == "SIM_WALK") {
                walk_start_time_ = now().seconds();
            }
            fsm_state_ = new_state;
        }
    }

    void loop_cb() {
        if (!session_) return;
        if (fsm_state_ != "WALK" && fsm_state_ != "SIM_WALK") return;
        if (joint_positions_.empty()) return;

        auto obs_array = obs_builder_.build(gyro_, gravity_, cmd_vel_, joint_positions_, joint_velocities_);
        // RCLCPP_INFO(get_logger(), "[RAW] cmd_vel=%.2f,%.2f,%.2f", cmd_vel_[0], cmd_vel_[1], cmd_vel_[2]);
        // RCLCPP_INFO(get_logger(), "[RAW] gyro=%.2f,%.2f,%.2f", gyro_[0], gyro_[1], gyro_[2]);
        // RCLCPP_INFO(get_logger(), "[RAW] gravity=%.2f,%.2f,%.2f", gravity_[0], gravity_[1], gravity_[2]);
        // RCLCPP_INFO(get_logger(), "[RAW] gravity=%.2f,%.2f,%.2f", gravity_[0], gravity_[1], gravity_[2]);
        RCLCPP_INFO(get_logger(), "[RAW] obs_array=%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,", obs_array[0],obs_array[1],obs_array[2],obs_array[3],
        obs_array[4],obs_array[5],obs_array[6],obs_array[7],obs_array[8],obs_array[9]*57.3,
        obs_array[10]*57.3,obs_array[11]*57.3,obs_array[12]*57.3,obs_array[13]*57.3,obs_array[14]*57.3,obs_array[15]*57.3,
        obs_array[16]*57.3,obs_array[17]*57.3,obs_array[18]*57.3,obs_array[19]*57.3,obs_array[20]*57.3,obs_array[21],
        obs_array[22],obs_array[23],obs_array[24],obs_array[25],obs_array[26],obs_array[27],
        obs_array[28],obs_array[29],obs_array[30],obs_array[31],obs_array[32],obs_array[33],
        obs_array[34],obs_array[35],obs_array[36],obs_array[37],obs_array[38],obs_array[39],
        obs_array[40],obs_array[41],obs_array[42],obs_array[43],obs_array[44]);
        // Run ONNX Inference
        Ort::MemoryInfo memory_info = Ort::MemoryInfo::CreateCpu(OrtArenaAllocator, OrtMemTypeDefault);
        std::vector<int64_t> input_shape = {1, 45};
        
        Ort::Value input_tensor = Ort::Value::CreateTensor<float>(
            memory_info, obs_array.data(), obs_array.size(), input_shape.data(), input_shape.size());

        auto output_tensors = session_->Run(
            Ort::RunOptions{nullptr}, 
            input_node_names_.data(), 
            &input_tensor, 1, 
            output_node_names_.data(), 1);

        float* out_arr = output_tensors.front().GetTensorMutableData<float>();
        std::array<float, 12> actions;
        for (int i = 0; i < 12; i++) {
            actions[i] = std::max(-1.0f, std::min(1.0f, out_arr[i]));
        }
        RCLCPP_INFO(get_logger(), "[RAW] actions=%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f", actions[0], actions[1], actions[2], actions[3], actions[4], actions[5], actions[6], actions[7], actions[8], actions[9], actions[10], actions[11]);
        obs_builder_.update_last_action(actions);

        debug_timer_++;
        int debug_freq = static_cast<int>(rate_) * 10;
        if (debug_timer_ % debug_freq == 1) {
            RCLCPP_INFO(get_logger(), "[RAW] cmd_vel=%.2f,%.2f,%.2f", cmd_vel_[0], cmd_vel_[1], cmd_vel_[2]);
        }

        auto targets = ObsBuilder::action_to_positions(actions);

        biped_msgs::msg::MITCommandArray cmd_msg;
        cmd_msg.header.stamp = now();

        double walk_ramp = 1.0;
        if (walk_start_time_ > 0.0) {
            double walk_elapsed = now().seconds() - walk_start_time_;
            walk_ramp = std::min(1.0, 0.1 + 0.9 * (walk_elapsed / WALK_GAIN_RAMP_SECS));
        }

        double gs = get_parameter("gain_scale").as_double();

        for (const auto& name : JOINT_ORDER) {
            auto kp_kd = gains_.at(name);
            biped_msgs::msg::MITCommand cmd;
            cmd.joint_name = name;
            cmd.position = targets.at(name);
            cmd.velocity = 0.0;
            cmd.kp = kp_kd.first * gs * walk_ramp;
            cmd.kd = kp_kd.second * gs * walk_ramp;
            cmd.torque_ff = 0.0;
            cmd_msg.commands.push_back(cmd);
        }

        if (fsm_state_ == "SIM_WALK") {
            pub_viz_->publish(cmd_msg);
            sensor_msgs::msg::JointState js;
            js.header = cmd_msg.header;
            for (const auto& cmd : cmd_msg.commands) {
                js.name.push_back(cmd.joint_name);
                js.position.push_back(cmd.position);
                js.velocity.push_back(0.0);
                js.effort.push_back(0.0);
            }
            pub_viz_js_->publish(js);
        } else if (fsm_state_ == "WALK") {
            pub_cmd_->publish(cmd_msg);
        }
    }
};

int main(int argc, char * argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<PolicyNodeCpp>());
    rclcpp::shutdown();
    return 0;
}
